x/airdrop/keeper/claim.go

```diff
@@ -0,0 +1,194 @@
+	if !userFound {
+		return types.ErrUserAllocationNotFound.Wrapf("user %s for airdrop %s", claimer, airdropId)
+	}
+
+	// Confirm the user has not elected to claim early (meaning their type will be ClaimDaily)
+	if userAllocation.ClaimType != types.CLAIM_DAILY {
+		return types.ErrClaimTypeUnavailable.Wrapf("user has already elected claim option %s",
+			userAllocation.ClaimType.String())
+	}
+
+	// Get the index in the allocations array from the current date
+	// E.g. on the 10th day of distribution, this will map to the 9th index in the list
+	windowLength := k.GetParams(ctx).AllocationWindowSeconds
+	todaysIndex, err := airdrop.GetCurrentDateIndex(ctx, windowLength)
+	if err != nil {
+		return err
+	}
+
+	// Sum the rewards up to that date and 0 them out in the process
+	todaysRewards := sdkmath.ZeroInt()
+	for i := 0; i <= todaysIndex; i++ {
+		rewardsOnDate := userAllocation.Allocations[i]
+		todaysRewards = todaysRewards.Add(rewardsOnDate)
+		userAllocation.Allocations[i] = sdkmath.ZeroInt()
+	}
+
+	// If there are no rewards, alert the user with an error
+	if todaysRewards.IsZero() {
+		return types.ErrNoUnclaimedRewards
+	}
+
+	// Update the amount claimed on the allocation record
+	userAllocation.Claimed = userAllocation.Claimed.Add(todaysRewards)
+
+	// Distribute rewards from the distributor
+	distributorAccount := sdk.MustAccAddressFromBech32(airdrop.DistributionAddress)
+	claimerAccount := sdk.MustAccAddressFromBech32(userAllocation.Address)
+	rewardsCoin := sdk.NewCoin(airdrop.RewardDenom, todaysRewards)
+
+	if err := k.bankKeeper.SendCoins(ctx, distributorAccount, claimerAccount, sdk.NewCoins(rewardsCoin)); err != nil {
+		return errorsmod.Wrapf(err, "unable to distribute rewards")
+	}
+
+	// Update the reward record for to mark the progress
+	k.SetUserAllocation(ctx, userAllocation)
+
+	return nil
+}
+
+// User transaction to claim a portion of their total amount now, and forfeit the
+// remainder to be clawed back
+func (k Keeper) ClaimEarly(ctx sdk.Context, airdropId, claimer string) error {
+	// Fetch the airdrop and user's allocations
+	airdrop, airdropFound := k.GetAirdrop(ctx, airdropId)
+	if !airdropFound {
+		return types.ErrAirdropNotFound.Wrapf("airdrop %s", airdropId)
+	}
+	userAllocation, userFound := k.GetUserAllocation(ctx, airdropId, claimer)
+	if !userFound {
+		return types.ErrUserAllocationNotFound.Wrapf("user %s for airdrop %s", claimer, airdropId)
+	}
+
+	// Confirm the user has not elected to claim early (meaning their type will be ClaimDaily)
+	if userAllocation.ClaimType != types.CLAIM_DAILY {
+		return types.ErrClaimTypeUnavailable.Wrapf("user has already elected claim option %s",
+			userAllocation.ClaimType.String())
+	}
+
+	// Confirm we're not past the decision date
+	currentTime := ctx.BlockTime().Unix()
+	if currentTime > airdrop.ClaimTypeDeadlineDate.Unix() {
+		return types.ErrAfterDecisionDeadline
+	}
+
+	// Confirm the airdrop started
+	if currentTime < airdrop.DistributionStartDate.Unix() {
```

Comment:
You're absolutely correct a92dcca01

Reply:
the reason it wasn't in `ClaimDaily` is because it's checked in `GetCurrentDateIndex`

Probably doesn't hurt to add a redundant check though for readability

Reply:
I haven't noticed it's already being checked in there, which makes me think "hiding" this in `GetCurrentDateIndex()` hurts readability and I personally prefer to double-check.

Reply:
lmk if you think otherwise

Reply:
yeah agreed!